cmd/stats-exporter/main.go

```diff
@@ -0,0 +1,114 @@
+package main
```

Comment:
Thanks for the PR @jkarner - A few top-level points of review feedback:

1) Can you expand on why you selected a design that builds both the `gzip` and `ssh` execution into the `stats-exporter`? I'm generally nervous to see Go code executing external binaries without strong justification, I think its a bit of an anti-pattern. It seems to me that the primary purpose of the `stats-exporter` is to export stats and the other concerns (compressing and copying) are separate. The operator could use their shell to perform compression and copying as required. As one example I think the default `-destination` argument being `"localhost:/tmp"` is a good sign that it isn't a given there's always a remote host that the data should be copied to.
2) There are no comments anywhere in `main.go` or `main_test.go`. I think at the minimum the types and functions should have short comments for godoc. The tests would probably benefit from some high level descriptions of what is being tested, and how. I think this would also help clarify the purpose of the `interface`'s used for mocking.
3) I'm a little bit nervous about a new Boulder binary being added that isn't used anywhere in CI or integration tests. We saw with `load-generator` and some other utilities that doing this tends to lead to bitrot that isn't caught until the tool is used by an operator. Is there any way we can run this tool as part of the integration tests to verify it doesn't rot?

Reply:
One other thought:
Is there a reason to prefer `gzip` instead of `bzip2` ? The latter will compress to smaller file sizes.

Reply:
1. This tool will either be executed daily as a cron job or a systemd timer. As an automated tool, I think it's best to have all the work done in one program that's easier to monitor and not need post execution logic and commands to make sure that the file was written correctly and sent off. If this was written as a shell script it would execute the mysql commands and then continue the work -- this is a better version of a simple script or program. 
2. Thank you for the reminder. I will add some comments
3. This new Boulder binary will only live in the letsencrypt/boulder repo for a short amount of time.The SRE team plans to set up a repo that has testing infrastructure and can house tools like this especially since it doesn't use the SA to make database queries. In the meantime, this tool will be running daily so any bitrot that could be incurred should be caught when the stats stop updating. More iterations on this tool could also include basic stats that indicate a success which could be monitored with prometheus.

`bzip2` isn't installed on the production OS. The daily export is relatively small and compression is more a nicety

Reply:
> 1. As an automated tool, I think it's best to have all the work done in one program that's easier to monitor and not need post execution logic and commands to make sure that the file was written correctly and sent off.

I think I understand your reasoning but for this instance I don't think a wrapper script adds much complexity and is the right tool for connecting up three independent tasks. The script would only be a few commands: one to run the stat exporter, one to gzip results, and one to scp the compressed data. There's nothing that would need to be checked except the exit code of each command. Your Go code is equivalent in that it doesn't check that the gzipped content was written correctly or that the SCP resulted in the right file in the right place - just that the exit codes were correct.

If consensus is that this tool should be written to avoid any external scripting I think one concrete improvement would be replacing the code that calls an external `gzip` process with code that uses the `compress/gzip` package from the Go stdlib to compress the data before writing it.

> 3. This new Boulder binary will only live in the letsencrypt/boulder repo for a short amount of time.The SRE team plans to set up a repo that has testing infrastructure and can house tools like this especially since it doesn't use the SA to make database queries. In the meantime, this tool will be running daily so any bitrot that could be incurred should be caught when the stats stop updating.

SGTM :+1: Thanks! Knowing that also makes me less reluctant to support building the `gzip` and `scp` into the command if that's what folks prefer. As a Boulder tool it should be as precisely focused on the Boulder specific problem as possible but as a SRE tool I can see wanting to fit it to your specific use case and environment.

> `bzip2` isn't installed on the production OS. The daily export is relatively small and compression is more a nicety

Ok, makes sense.

Reply:
> If consensus is that this tool should be written to avoid any external scripting

I think this is the general direction we want to go with most SRE tools. The goal is that most things we rely on should have unit or integration tests, and it's harder to write unittests for bash than it is for Python or Go. I agree that this style of doing things doesn't really match the rest of the Boulder's repo. I think Jillian's plan of eventually moving to a separate repo for SRE tools that has its own CI is a terrific one.

Reply:
> I think this is the general direction we want to go with most SRE tools. The goal is that most things we rely on should have unit or integration tests, and it's harder to write unittests for bash than it is for Python or Go.

While I mostly agree I think this is one of those places where it is a bit grey in terms of unit/integration testing. I think it makes sense to have Golang unit tests testing... well Golang logic, not external logic and integration tests testing the integration between different tooling. But ¯\_(ツ)_/¯, I don't think it's necessarily wrong to use Golang as the glue scripting language, it just feels a bit weird.

> I think Jillian's plan of eventually moving to a separate repo for SRE tools that has its own CI is a terrific one.

I agree this is a good idea, any reason not to start with this? We can still do reviews there and it feels a bit strange merging this into the tree if it's not _really_ a boulder tool. It doesn't feel like this is super urgent in terms of getting it into a tree and the initial setup of a SRE repo should be relatively lightweight.

Reply:
> I agree this is a good idea, any reason not to start with this? We can still do reviews there and it feels a bit strange merging this into the tree if it's not _really_ a boulder tool. It doesn't feel like this is super urgent in terms of getting it into a tree and the initial setup of a SRE repo should be relatively lightweight.

I agree it would be better to start in a separate repo right off the bat. For making a new repo extra fast I would recommend copying the [dns-lots-of-lookups .travis.yml](https://github.com/letsencrypt/dns-lots-of-lookups/blob/master/.travis.yml) and also adding https://golangci.com/ to the new repository along with Travis.

The few bits of `github.com/letsencrypt/boulder/cmd` being importing should also probably be copied into the new repo so there isn't a dependence on Boulder for a few tiny utility functions.

Reply:
> For making a new repo extra fast I would recommend copying the [dns-lots-of-lookups .travis.yml](https://github.com/letsencrypt/dns-lots-of-lookups/blob/master/.travis.yml) and also adding https://golangci.com/ to the new repository along with Travis.

I made a template repo that's even easier. If you want to follow this checklist and give me feedback it would be appreciated! https://github.com/letsencrypt/gorepotemplate/blob/master/CHECKLIST.md